tests/test_url.py

```diff
@@ -338,9 +338,20 @@ def test_ipv6_missing_right_bracket() -> None:
         URL("http://[1dec:0:0:0::1/")
 
 
-def test_ipv4_brackets_not_allowed() -> None:
-    with pytest.raises(ValueError, match="An IPv4 address cannot be in brackets"):
-        URL("http://[127.0.0.1]/")
+@pytest.mark.parametrize(
+    "url",
+    [
+        "http://[]/",  # Empty IPv6 URL
+        "http://[1]/",  # No semicolon
+        "http://[127.0.0.1]/",  # IPv4 inside brackets
+        "http://]1dec:0:0:0::1[/",  # Brackets in reversed order
+    ],
```

Comment:
Instead of having code comments, it's best to set explicit param ids. They will show up in the test report, and will make it easier to both reason about the testing semantics, plus can be used to select and run one of the test params with `pytest`:
```suggestion
    (
        "http://[]/",
        "http://[1]/",
        "http://[127.0.0.1]/",
        "http://]1dec:0:0:0::1[/",
    ),
    ids=(
        "empty-IPv6-like-URL",
        "no-colons-in-IPv6",
        "IPv4-inside-brackets",
        "brackets-in-reversed-order",
    ),
```